lib/AST/Type.cpp

```diff
@@ -1022,6 +1022,31 @@ void ProtocolType::canonicalizeProtocols(
   llvm::array_pod_sort(protocols.begin(), protocols.end(), compareProtocols);
 }
 
+static Type
+getCanonicalInputType(AnyFunctionType *funcType,
+                      std::function<CanType(Type)> getCanonicalType) {
+  auto inputType = funcType->getInput();
+  if (auto *TT = dyn_cast<TupleType>(inputType.getPointer()))
+    return getCanonicalType(TT);
+
+  if (auto *PT = dyn_cast<ParenType>(inputType.getPointer())) {
```

Comment:
Can this logic be simpler? It seems like we could do something like:

    bool isParen = isa<ParenType>(inputType.getPointer());
    inputType = inputType->getCanonicalType();
    if (isParen) inputType = ParenType::get(inputType);
    assert(AnyFunctionType::isCanonicalFunctionInputType(inputType));